lib/puppet/util.rb

```diff
@@ -25,20 +25,65 @@ module Util
 
   # Run some code with a specific environment.  Resets the environment back to
   # what it was at the end of the code.
-  def self.withenv(hash)
-    saved = ENV.to_hash
-    hash.each do |name, val|
-      ENV[name.to_s] = val
-    end
-
+  # Windows can store unicode chars in the environment, so need to use API calls
+  #   to get and set the environment instead of ENV due to encoding problems.
+  def self.withenv(hash, mode=:posix)
+    saved = get_environment(mode)
+    set_environment(hash, mode)
     yield
   ensure
-    ENV.clear
-    saved.each do |name, val|
-      ENV[name] = val
+    restore_environment(saved, mode) if saved
+  end
+
+  def self.get_environment(mode=:posix)
+    case mode
+      when :posix
+        ENV.to_hash
+      when :windows
+        require 'puppet/util/windows/process'
+        Puppet::Util::Windows::Process.get_environment_strings
+      else
+        raise "Unable to retrieve the environment for mode #{mode}"
     end
   end
 
+  def self.set_environment(hash, mode=:posix)
+    case mode
+      when :posix
+        hash.each do |name, val|
+          ENV[name.to_s] = val
+        end
+      when :windows
+        require 'puppet/util/windows/process'
+        hash.each do |name, val|
+          Puppet::Util::Windows::Process.set_environment_variable(name,val)
+        end
+      else
+        raise "Unable to set the environment for mode #{mode}"
+    end
+  end
+
+  def self.restore_environment(saved, mode=:posix)
```

Comment:
Also.. as an FYI, the code oesn't deal with case sensitivity properly.  In the default `ENV` for Windows, `ENV['foo'] == ENV['FoO']` (this is not true for POSIX which allows case sensitive variations on ENV var names).

So we'll need to add some code to deal with that.